### PR TITLE
[Clojure] Add fastText example

### PR DESCRIPTION
**Pull request by adc17**
## Description ##
- Right now the CNN text classification example provides support for glove and word2vec embeddings. This PR extends that support to fastText.
- Additionally, this PR updates the README.md instructions for running examples from the REPL.

Resolves https://github.com/apache/incubator-mxnet/issues/14118

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR).
- [x] For new examples, README.md is added to explain what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable.
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change.

### Changes ###
- [x] CNN text classification (including docs)

### CODE REVIEW

**Comment by gigasquid** [comment]
Great! I look forward to reviewing this 😸

**Comment by gigasquid** [comment]
pinging @Chouffe to take a look if he has time as well since he helped shape the issue ticket 😄

**Comment by Chouffe** [comment]
I will take a look shortly @adc17 
Looking forward to reviewing it!

**Comment by Chouffe** [comment]
I could run fastText:
```clojure
cnn-text-classification.classifier=> (train-convnet {:devs [(context/cpu 0)] :embedding-size 300 :batch-size 100 :test-size 100 :num-epoch 10 :max-examples 1000 :pretrained-embedding :fastText})
Loading all the movie reviews from  data/mr-data
WARN  org.apache.mxnet.WarnIfNotDisposed: LEAK: [one-time warning] An instance of org.apache.mxnet.Symbol was not disposed. Set property mxnet.traceLeakedObjects to true to enable tracing
Loading the fastText pre-trained word embeddings from  data/fastText/wiki.simple.vec
Shuffling the data and splitting into training and test sets
{:sentence-count 2000, :sentence-size 62, :vocab-size 8078, :embedding-size 300, :pretrained-embedding :fastText}
Getting ready to train for  10  epochs
===========
WARN  org.apache.mxnet.DataDesc: Found Undefined Layout, will use default index 0 for batch axis
WARN  org.apache.mxnet.DataDesc: Found Undefined Layout, will use default index 0 for batch axis
WARN  org.apache.mxnet.DataDesc: Found Undefined Layout, will use default index 0 for batch axis
WARN  org.apache.mxnet.DataDesc: Found Undefined Layout, will use default index 0 for batch axis
WARN  org.apache.mxnet.DataDesc: Found Undefined Layout, will use default index 0 for batch axis
[18:54:04] src/operator/tensor/./matrix_op-inl.h:200: Using target_shape will be deprecated.
[18:54:04] src/operator/tensor/./matrix_op-inl.h:200: Using target_shape will be deprecated.
INFO  org.apache.mxnet.module.BaseModule: Epoch[0] Train-accuracy=0.5326316
INFO  org.apache.mxnet.module.BaseModule: Epoch[0] Time cost=4463
INFO  org.apache.mxnet.module.BaseModule: Epoch[0] Validation-accuracy=0.59
...
INFO  org.apache.mxnet.module.BaseModule: Epoch[8] Train-accuracy=0.9836842
INFO  org.apache.mxnet.module.BaseModule: Epoch[8] Time cost=4093
INFO  org.apache.mxnet.module.BaseModule: Epoch[8] Validation-accuracy=0.73
INFO  org.apache.mxnet.module.BaseModule: Epoch[9] Train-accuracy=0.9878947
INFO  org.apache.mxnet.module.BaseModule: Epoch[9] Time cost=3861
INFO  org.apache.mxnet.module.BaseModule: Epoch[9] Validation-accuracy=0.75
```

Thanks a lot for adding this @adc17! It seems to work really well :-)
I am wondering if we should add bash scripts to fetch the data (for word2vec and fastText embeddings)?

File: contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj [hunk 1]
```diff
@@ -90,15 +92,21 @@
...
+(defn read-text-embedding-pairs [pairs]
+  (for [^String line pairs
         :let [fields (.split line " ")]]
     [(aget fields 0)
      (mapv #(Float/parseFloat ^String %) (rest fields))]))
 
 (defn load-glove [glove-file-path]
   (println "Loading the glove pre-trained word embeddings from " glove-file-path)
-  (into {} (read-text-embedding-pairs (io/reader glove-file-path))))
+  (into {} (read-text-embedding-pairs (line-seq (io/reader glove-file-path)))))
```
Comment (Chouffe) [inline contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj]: It is becoming hard to read. Can we use a threading macro here?
```clojure
(->> (io/reader path)
     line-seq
     read-text-embedding-pairs
     (into {}))
```

File: contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj [hunk 2]
```diff
@@ -90,15 +92,21 @@
...
+
+(def remove-fasttext-metadata rest)
+
+(defn load-fasttext [fasttext-file-path]
+  (println "Loading the fastText pre-trained word embeddings from " fasttext-file-path)
+  (into {} (read-text-embedding-pairs (remove-fasttext-metadata (line-seq (io/reader fasttext-file-path))))))
```
Comment (Chouffe) [inline contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj]: Can we also use a threading macro here for readability?

File: contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj [hunk 3]
```diff
@@ -190,6 +198,8 @@
         vocab-embeddings (case pretrained-embedding
                            :glove (->> (load-glove (glove-file-path embedding-size))
                                        (build-vocab-embeddings vocab embedding-size))
+                           :fastText (->> (load-fasttext fasttext-file-path)
```
Comment (Chouffe) [inline contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj]: I would prefer a keyword like `:fast-text` or `:fasttext` instead of having an uppercase character.

File: contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj [hunk 4]
```diff
@@ -90,15 +92,21 @@
...
      (mapv #(Float/parseFloat ^String %) (rest fields))]))
 
 (defn load-glove [glove-file-path]
   (println "Loading the glove pre-trained word embeddings from " glove-file-path)
-  (into {} (read-text-embedding-pairs (io/reader glove-file-path))))
+  (into {} (read-text-embedding-pairs (line-seq (io/reader glove-file-path)))))
+
+(def remove-fasttext-metadata rest)
+
+(defn load-fasttext [fasttext-file-path]
```
Comment (Chouffe) [inline contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj]: Should we mark the IO functions with `!` to convey it?
```clojure
(defn load-fasttext! ...)
```

File: contrib/clojure-package/examples/cnn-text-classification/README.md [hunk 5]
```diff
@@ -58,15 +72,15 @@ you'll need to unzip them and place them in the `contrib/clojure-package/data` d
...
+(train-convnet {:devs [(context/cpu 0)] :embedding-size 300 :batch-size 100 :test-size 100 :num-epoch 10 :max-examples 1000 :pretrained-embedding :word2vec})
 ```
 Note that loading word2vec embeddings consumes memory and takes some time.
 
 You can also train them using `JVM_OPTS="-Xmx8g" lein run` once you've modified
 the parameters to `train-convnet` (see above) in `src/cnn_text_classification/classifier.clj`.
 In order to run training with word2vec on the complete data set, you will need to run:
 ```
-(train-convnet {:embedding-size 300 :batch-size 100 :test-size 1000 :num-epoch 10 :pretrained-embedding :word2vec})
+(train-convnet {:devs [(context/cpu 0)] :embedding-size 300 :batch-size 100 :test-size 1000 :num-epoch 10 :pretrained-embedding :word2vec})
```
Comment (Chouffe) [inline contrib/clojure-package/examples/cnn-text-classification/README.md]: Thanks for fixing the README!

File: contrib/clojure-package/examples/cnn-text-classification/README.md [hunk 6]
```diff
@@ -58,15 +72,15 @@ you'll need to unzip them and place them in the `contrib/clojure-package/data` d
 
 Then you can run training on a subset of examples through the repl using:
 ```
-(train-convnet {:embedding-size 300 :batch-size 100 :test-size 100 :num-epoch 10 :max-examples 1000 :pretrained-embedding :word2vec})
+(train-convnet {:devs [(context/cpu 0)] :embedding-size 300 :batch-size 100 :test-size 100 :num-epoch 10 :max-examples 1000 :pretrained-embedding :word2vec})
```
Comment (Chouffe) [inline contrib/clojure-package/examples/cnn-text-classification/README.md]: Thanks for fixing the README! Should we use `(context/default-context)` instead though?

File: contrib/clojure-package/examples/cnn-text-classification/README.md [hunk 7]
```diff
@@ -49,6 +48,21 @@ and then run
...
+
+Using fastText instead of glove is fairly straightforward, as the pretrained embedding format is very similar.
+
+Download the 'Simple English' pretrained wiki word vectors (text) from the fastText
+[site](https://fasttext.cc/docs/en/pretrained-vectors.html) and place them in the
+`data/fastText` directory.
+
+Then you can run training on a subset of examples through the repl using:
+```
+(train-convnet {:devs [(context/cpu 0)] :embedding-size 300 :batch-size 100 :test-size 100 :num-epoch 10 :max-examples 1000 :pretrained-embedding :fastText})
```
Comment (Chouffe) [inline contrib/clojure-package/examples/cnn-text-classification/README.md]: Should we use `(context/default-context)` instead?

File: contrib/clojure-package/examples/cnn-text-classification/README.md [hunk 8]
```diff
@@ -29,8 +29,7 @@ You also must download the glove word embeddings. The suggested one to use is th
 ## Usage
 
 You can run through the repl with
-`(train-convnet {:embedding-size 50 :batch-size 100 :test-size 100 :num-epoch 10 :max-examples 1000 :pretrained-embedding :glove})`
-
+`(train-convnet {:devs [(context/cpu 0)] :embedding-size 50 :batch-size 100 :test-size 100 :num-epoch 10 :max-examples 1000 :pretrained-embedding :glove})`
```
Comment (Chouffe) [inline contrib/clojure-package/examples/cnn-text-classification/README.md]: Thanks for fixing the README! Should we use `(context/default-context)` instead?

**Comment by gigasquid** [comment]
Thanks for your contribution @adc17 and thanks for reviewing @Chouffe 😸 

I took a look and it looks good to me too. When then feedback items are addressed and CI is green, it will be good to merge.

File: contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj (hunk 1 above)
Comment (adc17) [reply contrib/clojure-package/examples/cnn-text-classification/src/cnn_text_classification/data_helper.clj]: Yep, I considered this myself!

**Comment by adc17** [comment]
Thanks for the thorough review @Chouffe; I agree with all your suggestions.
